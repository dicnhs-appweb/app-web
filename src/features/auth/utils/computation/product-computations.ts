import {
  OverheadExpense,
  ProfitMarginSettings,
  RawMaterial,
} from '@/types/product.schema';
import Decimal from 'decimal.js';
import {
  calculateTotalOverheadCost,
  calculateTotalRawMaterialCost,
} from './sub-computations';

function calculateManufacturingCostPerUnit(
  rawMaterials: RawMaterial[],
  overheadExpenses: OverheadExpense[]
): number {
  const rawMaterialCost = new Decimal(
    calculateTotalRawMaterialCost(rawMaterials)
  );
  const overheadCost = new Decimal(
    calculateTotalOverheadCost(overheadExpenses)
  );
  return rawMaterialCost.plus(overheadCost).toDecimalPlaces(2).toNumber();
}

function calculateRecommendedRetailPrice(
  manufacturingCostPerUnit: number,
  profitMarginSettings: ProfitMarginSettings
): number {
  const {calculationType, profitValue} = profitMarginSettings;
  const cost = new Decimal(manufacturingCostPerUnit);
  const profit = new Decimal(profitValue);

  const price =
    calculationType === 'fixed-markup'
      ? cost.plus(profit)
      : cost.times(Decimal.sum(1, profit.dividedBy(100)));

  return price.toDecimalPlaces(2).toNumber();
}

function calculateBreakEvenPoint(
  overheadExpenses: OverheadExpense[],
  rawMaterials: RawMaterial[],
  recommendedRetailPrice: number
): number {
  const fixedCosts = new Decimal(calculateTotalOverheadCost(overheadExpenses));
  const variableCostPerUnit = new Decimal(
    calculateTotalRawMaterialCost(rawMaterials)
  );
  const contributionMargin = new Decimal(recommendedRetailPrice).minus(
    variableCostPerUnit
  );

  return fixedCosts.dividedBy(contributionMargin).toDecimalPlaces(2).toNumber();
}

function calculateProfitPerUnit(
  recommendedRetailPrice: number,
  manufacturingCostPerUnit: number
): number {
  return new Decimal(recommendedRetailPrice)
    .minus(manufacturingCostPerUnit)
    .toDecimalPlaces(2)
    .toNumber();
}

function calculateTotalPotentialProfit(
  profitPerUnit: number,
  desiredProductionQuantity: number
): number {
  return new Decimal(profitPerUnit)
    .times(desiredProductionQuantity)
    .toDecimalPlaces(2)
    .toNumber();
}

function calculateMarginOfSafety(
  expectedSalesQuantity: number,
  breakEvenPoint: number
): number {
  if (expectedSalesQuantity <= 0) {
    return 0;
  }
  return new Decimal(expectedSalesQuantity)
    .minus(breakEvenPoint)
    .dividedBy(expectedSalesQuantity)
    .times(100)
    .toDecimalPlaces(2)
    .toNumber();
}

function calculateContributionMargin(
  recommendedRetailPrice: number,
  rawMaterials: RawMaterial[]
): number {
  const variableCostPerUnit = new Decimal(
    calculateTotalRawMaterialCost(rawMaterials)
  );
  return new Decimal(recommendedRetailPrice)
    .minus(variableCostPerUnit)
    .toDecimalPlaces(2)
    .toNumber();
}

export {
  calculateBreakEvenPoint,
  calculateContributionMargin,
  calculateManufacturingCostPerUnit,
  calculateMarginOfSafety,
  calculateProfitPerUnit,
  calculateRecommendedRetailPrice,
  calculateTotalPotentialProfit,
};
