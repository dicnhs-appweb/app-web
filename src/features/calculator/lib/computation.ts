import {
  CostInfo,
  DirectCost,
  FinancialAnalysisResult,
  IndirectCost,
  PricingStrategy,
  ProductPricingModel,
  ProfitScenario,
} from '../types/product.model';

import Decimal from 'decimal.js';
import memoize from 'lodash/memoize';

const calculateTotalCost = (costs: CostInfo[]): number => {
  return costs
    .reduce(
      (total, cost) =>
        total.plus(new Decimal(cost.quantity).times(cost.unitCost)),
      new Decimal(0)
    )
    .toNumber();
};

export function calculateUnitCost(
  directCosts: DirectCost[],
  indirectCosts: IndirectCost[],
  plannedProductionQuantity: number
): number {
  const totalDirectCost = calculateTotalCost(directCosts);
  const totalIndirectCost = calculateTotalCost(indirectCosts);
  const unitDirectCost = new Decimal(totalDirectCost)
    .dividedBy(plannedProductionQuantity)
    .toDecimalPlaces(2)
    .toNumber();
  const unitIndirectCost = new Decimal(totalIndirectCost)
    .dividedBy(plannedProductionQuantity)
    .toDecimalPlaces(2)
    .toNumber();
  const totalUnitCost = new Decimal(unitDirectCost)
    .plus(unitIndirectCost)
    .toDecimalPlaces(2)
    .toNumber();

  return totalUnitCost;
}

export function calculateRecommendedPrice(
  unitCost: number,
  pricingStrategy: PricingStrategy
): number {
  const cost = new Decimal(unitCost);
  const targetValue = new Decimal(pricingStrategy.targetValue);

  const price =
    pricingStrategy.method === 'cost-plus'
      ? cost.times(Decimal.sum(1, targetValue.dividedBy(100)))
      : cost.plus(targetValue);

  return price.toDecimalPlaces(2).toNumber();
}

export function calculateBreakEvenPoint(
  totalIndirectCost: number,
  unitPrice: number,
  unitVariableCost: number
): number {
  const contributionMarginPerUnit = new Decimal(unitPrice).minus(
    unitVariableCost
  );
  if (contributionMarginPerUnit.isZero()) {
    throw new Error('Contribution margin per unit cannot be zero');
  }
  return new Decimal(totalIndirectCost)
    .dividedBy(contributionMarginPerUnit)
    .ceil()
    .toNumber();
}

export function calculateContributionMargin(
  unitPrice: number,
  unitVariableCost: number
): number {
  return new Decimal(unitPrice)
    .minus(unitVariableCost)
    .toDecimalPlaces(2)
    .toNumber();
}

export function calculateContributionMarginRatio(
  contributionMargin: number,
  unitPrice: number
): number {
  return new Decimal(contributionMargin)
    .dividedBy(unitPrice)
    .toDecimalPlaces(4)
    .toNumber();
}

export function calculateOperatingLeverage(
  quantity: number,
  contributionMargin: number,
  fixedOperatingCosts: number
): number {
  const totalContributionMargin = new Decimal(quantity).times(
    contributionMargin
  );
  if (totalContributionMargin.equals(fixedOperatingCosts)) {
    return 0;
  }
  return totalContributionMargin
    .dividedBy(totalContributionMargin.minus(fixedOperatingCosts))
    .abs()
    .toDecimalPlaces(2)
    .toNumber();
}

export const calculateProfitScenarios = memoize(
  (
    fixedCosts: number,
    unitVariableCost: number,
    unitPrice: number,
    maxQuantity: number
  ): ProfitScenario[] => {
    const scenarios: ProfitScenario[] = [];
    const minStepSize = 1;
    const maxStepSize = Math.max(1, Math.floor(maxQuantity / 50));
    let stepSize = minStepSize;

    for (
      let salesVolume = 0;
      salesVolume <= maxQuantity;
      salesVolume += stepSize
    ) {
      const totalRevenue = new Decimal(unitPrice).times(salesVolume);
      const totalCost = new Decimal(fixedCosts).plus(
        new Decimal(unitVariableCost).times(salesVolume)
      );
      const profit = totalRevenue.minus(totalCost);
      scenarios.push({salesVolume, profit: profit.toNumber()});

      // Adjust step size dynamically
      if (scenarios.length > 50) {
        stepSize = Math.min(stepSize * 2, maxStepSize);
      }
    }

    // Ensure the last point is always included
    if (scenarios[scenarios.length - 1].salesVolume !== maxQuantity) {
      const totalRevenue = new Decimal(unitPrice).times(maxQuantity);
      const totalCost = new Decimal(fixedCosts).plus(
        new Decimal(unitVariableCost).times(maxQuantity)
      );
      const profit = totalRevenue.minus(totalCost);
      scenarios.push({salesVolume: maxQuantity, profit: profit.toNumber()});
    }

    return scenarios;
  },
  (...args: [number, number, number, number]) => JSON.stringify(args)
);

export function performFinancialAnalysis(
  product: ProductPricingModel
): FinancialAnalysisResult | undefined {
  const result = ProductPricingModel.safeParse(product);

  if (!result.success) {
    return undefined;
  }

  const unitCost = calculateUnitCost(
    product.directCosts || [],
    product.indirectCosts || [],
    product.plannedProductionQuantity
  );

  const recommendedPrice = calculateRecommendedPrice(
    unitCost,
    product.pricingStrategy
  );

  const totalDirectCost = calculateTotalCost(product.directCosts || []);
  const totalIndirectCost = calculateTotalCost(product.indirectCosts || []);

  const unitVariableCost = new Decimal(totalDirectCost)
    .dividedBy(product.plannedProductionQuantity)
    .toNumber();

  const breakEvenPoint = calculateBreakEvenPoint(
    totalIndirectCost,
    recommendedPrice,
    unitVariableCost
  );

  const contributionMargin = calculateContributionMargin(
    recommendedPrice,
    unitVariableCost
  );
  const contributionMarginRatio = calculateContributionMarginRatio(
    contributionMargin,
    recommendedPrice
  );

  const totalCost = new Decimal(totalIndirectCost)
    .plus(new Decimal(unitVariableCost).times(product.forecastedSalesQuantity))
    .toNumber();

  const operatingLeverage = calculateOperatingLeverage(
    product.forecastedSalesQuantity,
    contributionMargin,
    totalIndirectCost
  );

  const profitScenarios = calculateProfitScenarios(
    totalIndirectCost,
    unitVariableCost,
    recommendedPrice,
    Math.max(product.plannedProductionQuantity, product.forecastedSalesQuantity)
  );

  return {
    // cost-related metrics
    unitCost,
    totalCost,
    totalDirectCost,
    totalIndirectCost,
    // pricing metrics
    recommendedPrice,
    // break-even analysis
    breakEvenPoint,
    // profitability metrics
    contributionMargin,
    contributionMarginRatio,
    // risk and sensitivity metrics
    operatingLeverage,
    // scenario analysis
    profitScenarios,
  };
}
