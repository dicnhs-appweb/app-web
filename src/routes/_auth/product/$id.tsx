import {productByIdQueryOptions} from '@/features/auth/dashboard/api/get-product'
import {ProductType} from '@/types/product.type'
import {useSuspenseQuery} from '@tanstack/react-query'
import {createFileRoute} from '@tanstack/react-router'
import {z} from 'zod'

export const Route = createFileRoute('/_auth/product/$id')({
  params: {
    parse: params => ({
      id: z
        .string()
        .regex(/^[0-9a-fA-F]{24}$/)
        .parse(params.id),
    }),
    stringify: ({id}) => ({id}),
  },
  loader: opts =>
    opts.context.queryClient.ensureQueryData(
      productByIdQueryOptions(opts.params.id)
    ),
  component: ViewProductComponent,
})

function ViewProductComponent() {
  const {id} = Route.useParams()
  const productQuery = useSuspenseQuery(productByIdQueryOptions(id))
  const product = productQuery.data as ProductType

  return (
    <div className="container mx-auto p-4">
      <h1 className="text-2xl font-bold mb-4">Product Details</h1>
      <div className="bg-white shadow-md rounded px-8 pt-6 pb-8 mb-4">
        <div className="mb-4">
          <label className="block text-gray-700 text-sm font-bold mb-2">
            Product Name:
          </label>
          <p className="text-gray-900">{product.productName}</p>
        </div>
        <div className="mb-4">
          <label className="block text-gray-700 text-sm font-bold mb-2">
            Manufacturing Cost Per Unit:
          </label>
          <p className="text-gray-900">
            ${product.manufacturingCostPerUnit?.toFixed(2) ?? 'N/A'}
          </p>
        </div>
        <div className="mb-4">
          <label className="block text-gray-700 text-sm font-bold mb-2">
            Profit Margin Settings:
          </label>
          <p className="text-gray-900">
            Type: {product.profitMarginSettings.calculationType}, Value:{' '}
            {product.profitMarginSettings.percentageValue}%
          </p>
        </div>
        <div className="mb-4">
          <label className="block text-gray-700 text-sm font-bold mb-2">
            Recommended Retail Price:
          </label>
          <p className="text-gray-900">
            ${product.recommendedRetailPrice?.toFixed(2) ?? 'N/A'}
          </p>
        </div>
        <div className="mb-4">
          <label className="block text-gray-700 text-sm font-bold mb-2">
            Production Forecast:
          </label>
          <p className="text-gray-900">
            Max Units:{' '}
            {product.productionForecast?.maximumUnitsProducible ?? 'N/A'},
            Constraint:{' '}
            {product.productionForecast?.productionConstraint || 'None'}
          </p>
        </div>
        <div className="mb-4">
          <label className="block text-gray-700 text-sm font-bold mb-2">
            Raw Materials:
          </label>
          <ul className="list-disc pl-5">
            {product.rawMaterials.map((material, index) => (
              <li key={index} className="mb-2">
                {material.ingredientName}: {material.quantityNeededPerUnit}{' '}
                units needed, ${material.costPerUnit.toFixed(2)} per unit,
                {material.stockOnHand} in stock
              </li>
            ))}
          </ul>
        </div>
        <div className="mb-4">
          <label className="block text-gray-700 text-sm font-bold mb-2">
            Overhead Expenses:
          </label>
          <ul className="list-disc pl-5">
            {product.overheadExpenses.map((expense, index) => (
              <li key={index}>
                {expense.expenseCategory}: ${expense.costPerUnit.toFixed(2)} per
                unit
              </li>
            ))}
          </ul>
        </div>
      </div>
    </div>
  )
}

export default ViewProductComponent
