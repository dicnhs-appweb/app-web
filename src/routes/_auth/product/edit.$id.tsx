import {createFileRoute} from '@tanstack/react-router'

export const Route = createFileRoute('/_auth/product/edit/$id')({
  component: () => <div>Hello /_auth/product/edit/$id!</div>,
})
