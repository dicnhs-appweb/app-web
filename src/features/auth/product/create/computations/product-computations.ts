import Decimal from 'decimal.js';
import {
  calculateTotalOverheadCost,
  calculateTotalRawMaterialCost,
} from './lib/get-total';
import {
  calculateBreakEvenPoint,
  calculateContributionMargin,
  calculateManufacturingCostPerUnit,
  calculateMarginOfSafety,
  calculateProfitPerUnit,
  calculateProfitRange,
  calculateRecommendedSalesPrice,
  calculateTotalPotentialProfit,
} from './product-formula';
import {ComputationResult, ProductSchema} from './types/product.schema';

export const getComputationResult = (
  product: ProductSchema
): ComputationResult => {
  const {
    rawMaterials,
    overheadExpenses,
    profitMarginSettings,
    expectedSales,
    quantityProduced,
  } = product;
  const manufacturingCostPerUnit = calculateManufacturingCostPerUnit(
    rawMaterials ?? [],
    overheadExpenses ?? [],
    quantityProduced
  );
  const recommendedSalesPrice = calculateRecommendedSalesPrice(
    manufacturingCostPerUnit,
    profitMarginSettings
  );
  const breakEvenPoint = calculateBreakEvenPoint(
    overheadExpenses ?? [],
    rawMaterials ?? [],
    recommendedSalesPrice,
    quantityProduced
  );
  const profitPerUnit = calculateProfitPerUnit(
    recommendedSalesPrice,
    manufacturingCostPerUnit
  );
  const marginOfSafety = calculateMarginOfSafety(expectedSales, breakEvenPoint);
  const totalPotentialProfit = calculateTotalPotentialProfit(
    profitPerUnit,
    expectedSales
  );
  const contributionMargin = calculateContributionMargin(
    recommendedSalesPrice,
    rawMaterials ?? [],
    quantityProduced
  );

  const fixedCosts = calculateTotalOverheadCost(overheadExpenses ?? []);
  const variableCostPerUnit = new Decimal(
    calculateTotalRawMaterialCost(rawMaterials ?? [])
  )
    .dividedBy(quantityProduced)
    .toNumber();
  const profitRange = calculateProfitRange(
    fixedCosts,
    variableCostPerUnit,
    recommendedSalesPrice,
    quantityProduced
  );

  const totalManufacturingCost = new Decimal(manufacturingCostPerUnit)
    .times(quantityProduced)
    .toNumber();

  return {
    manufacturingCostPerUnit,
    recommendedSalesPrice,
    breakEvenPoint,
    profitPerUnit,
    totalPotentialProfit,
    marginOfSafety,
    contributionMargin,
    profitRange,
    totalManufacturingCost,
  };
};
