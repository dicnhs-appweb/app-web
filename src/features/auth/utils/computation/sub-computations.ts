import {OverheadExpense, RawMaterial} from '@/types/product.schema';
import Decimal from 'decimal.js';

function calculateTotalRawMaterialCost(rawMaterials: RawMaterial[]): number {
  if (!Array.isArray(rawMaterials) || rawMaterials.length === 0) {
    return 0;
  }
  let totalCost = new Decimal(0);
  for (const {costPerUnit, totalUnits} of rawMaterials) {
    totalCost = totalCost.plus(new Decimal(costPerUnit).times(totalUnits));
  }
  return totalCost.toDecimalPlaces(2).toNumber();
}

function calculateTotalOverheadCost(
  overheadExpenses: OverheadExpense[]
): number {
  if (!Array.isArray(overheadExpenses) || overheadExpenses.length === 0) {
    return 0;
  }
  let totalCost = new Decimal(0);
  for (const {costPerUnit, totalUnits} of overheadExpenses) {
    totalCost = totalCost.plus(new Decimal(costPerUnit).times(totalUnits));
  }
  return totalCost.toDecimalPlaces(2).toNumber();
}

export {calculateTotalOverheadCost, calculateTotalRawMaterialCost};
