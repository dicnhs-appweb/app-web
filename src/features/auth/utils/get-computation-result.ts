import {ComputationResult, ProductSchema} from '@/types/product.schema';
import {
  calculateBreakEvenPoint,
  calculateContributionMargin,
  calculateManufacturingCostPerUnit,
  calculateMarginOfSafety,
  calculateProfitPerUnit,
  calculateRecommendedRetailPrice,
  calculateTotalPotentialProfit,
} from './computation/product-computations';

export const getComputationResult = (
  product: ProductSchema
): ComputationResult => {
  const {rawMaterials, overheadExpenses, profitMarginSettings, expectedSales} =
    product;

  const manufacturingCostPerUnit = calculateManufacturingCostPerUnit(
    rawMaterials,
    overheadExpenses
  );
  const recommendedRetailPrice = calculateRecommendedRetailPrice(
    manufacturingCostPerUnit,
    profitMarginSettings
  );
  const breakEvenPoint = calculateBreakEvenPoint(
    overheadExpenses,
    rawMaterials,
    recommendedRetailPrice
  );
  const profitPerUnit = calculateProfitPerUnit(
    recommendedRetailPrice,
    manufacturingCostPerUnit
  );
  const marginOfSafety = calculateMarginOfSafety(expectedSales, breakEvenPoint);
  const totalPotentialProfit = calculateTotalPotentialProfit(
    profitPerUnit,
    expectedSales
  );
  const contributionMargin = calculateContributionMargin(
    recommendedRetailPrice,
    rawMaterials
  );

  return {
    manufacturingCostPerUnit,
    recommendedRetailPrice,
    breakEvenPoint,
    profitPerUnit,
    totalPotentialProfit,
    marginOfSafety,
    contributionMargin,
  };
};
