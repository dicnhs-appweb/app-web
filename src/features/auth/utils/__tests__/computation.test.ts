import {ProductSchema} from '@/types/product.schema';
import * as productComputations from '../computation/product-computations';
import {getComputationResult} from '../get-computation-result';

jest.mock('../computation/sub-computations', () => ({
  calculateTotalOverheadCost: jest.fn().mockReturnValue(50),
  calculateTotalRawMaterialCost: jest.fn().mockReturnValue(100),
}));

const mockData = {
  basicProduct: {
    productName: 'Basic Product',
    profitMarginSettings: {
      calculationType:
        'fixed-markup' as ProductSchema['profitMarginSettings']['calculationType'],
      profitValue: 10,
    },
    desiredProductionQuantity: 1,
    expectedSales: 1,
    rawMaterials: [
      {materialName: 'Material A', totalUnits: 1, costPerUnit: 5},
      {materialName: 'Material B', totalUnits: 1, costPerUnit: 5},
    ],
    overheadExpenses: [{expenseName: 'Labor', totalUnits: 1, costPerUnit: 10}],
  },
  complexProduct: {
    productName: 'Complex Product',
    profitMarginSettings: {
      calculationType: 'fixed-markup' as const,
      profitValue: 50,
    },
    desiredProductionQuantity: 1000,
    expectedSales: 800,
    rawMaterials: [
      {materialName: 'Material X', totalUnits: 500, costPerUnit: 2.75},
      {materialName: 'Material Y', totalUnits: 200, costPerUnit: 5.5},
      {materialName: 'Material Z', totalUnits: 100, costPerUnit: 10.25},
    ],
    overheadExpenses: [
      {expenseName: 'Rent', totalUnits: 1, costPerUnit: 2000},
      {expenseName: 'Labor', totalUnits: 160, costPerUnit: 15},
      {expenseName: 'Utilities', totalUnits: 1, costPerUnit: 500},
    ],
  },
  edgeCaseProduct: {
    productName: 'Edge Case Product',
    profitMarginSettings: {
      calculationType: 'cost-plus' as const,
      profitValue: -10,
    },
    desiredProductionQuantity: 0,
    expectedSales: 0,
    rawMaterials: [],
    overheadExpenses: [],
  },
  noOverheadProduct: {
    productName: 'No Overhead Product',
    profitMarginSettings: {
      calculationType: 'fixed-markup' as const,
      profitValue: 20,
    },
    desiredProductionQuantity: 100,
    expectedSales: 80,
    rawMaterials: [
      {materialName: 'Material A', totalUnits: 100, costPerUnit: 5},
    ],
    overheadExpenses: [],
  },
  noRawMaterialProduct: {
    productName: 'No Raw Material Product',
    profitMarginSettings: {
      calculationType: 'fixed-markup' as const,
      profitValue: 30,
    },
    desiredProductionQuantity: 50,
    expectedSales: 40,
    rawMaterials: [],
    overheadExpenses: [{expenseName: 'Rent', totalUnits: 1, costPerUnit: 1000}],
  },
  highProfitMarginProduct: {
    productName: 'High Profit Margin Product',
    profitMarginSettings: {
      calculationType: 'fixed-markup' as const,
      profitValue: 1000,
    },
    desiredProductionQuantity: 10,
    expectedSales: 8,
    rawMaterials: [
      {materialName: 'Material A', totalUnits: 10, costPerUnit: 10},
    ],
    overheadExpenses: [{expenseName: 'Labor', totalUnits: 1, costPerUnit: 50}],
  },
  costPlusProduct: {
    productName: 'Cost Plus Product',
    profitMarginSettings: {
      calculationType: 'cost-plus' as const,
      profitValue: 25,
    },
    desiredProductionQuantity: 500,
    expectedSales: 400,
    rawMaterials: [
      {materialName: 'Material A', totalUnits: 500, costPerUnit: 2},
    ],
    overheadExpenses: [
      {expenseName: 'Labor', totalUnits: 100, costPerUnit: 10},
    ],
  },
};

const expectedResults = {
  basicProduct: {
    manufacturingCostPerUnit: 20,
    recommendedRetailPrice: 30,
    breakEvenPoint: 0.5,
    profitPerUnit: 10,
    totalPotentialProfit: 10,
    marginOfSafety: 50,
    contributionMargin: 20,
  },
  complexProduct: {
    manufacturingCostPerUnit: 5400,
    recommendedRetailPrice: 5450,
    breakEvenPoint: 0.99,
    profitPerUnit: 50,
    totalPotentialProfit: 50000,
    marginOfSafety: -19.95,
    contributionMargin: 4350,
  },
  edgeCaseProduct: {
    manufacturingCostPerUnit: 0,
    recommendedRetailPrice: 0,
    breakEvenPoint: 0,
    profitPerUnit: 0,
    totalPotentialProfit: 0,
    marginOfSafety: 0,
    contributionMargin: 0,
  },
  noOverheadProduct: {
    manufacturingCostPerUnit: 500,
    recommendedRetailPrice: 520,
    breakEvenPoint: 0,
    profitPerUnit: 20,
    totalPotentialProfit: 2000,
    marginOfSafety: 100,
    contributionMargin: 520,
  },
  noRawMaterialProduct: {
    manufacturingCostPerUnit: 1000,
    recommendedRetailPrice: 1030,
    breakEvenPoint: 33.33,
    profitPerUnit: 30,
    totalPotentialProfit: 1500,
    marginOfSafety: 16.67,
    contributionMargin: 1030,
  },
  highProfitMarginProduct: {
    manufacturingCostPerUnit: 150,
    recommendedRetailPrice: 1150,
    breakEvenPoint: 0.05,
    profitPerUnit: 1000,
    totalPotentialProfit: 10000,
    marginOfSafety: 99.94,
    contributionMargin: 1100,
  },
  costPlusProduct: {
    manufacturingCostPerUnit: 2000,
    recommendedRetailPrice: 2500,
    breakEvenPoint: 400,
    profitPerUnit: 500,
    totalPotentialProfit: 250000,
    marginOfSafety: 0,
    contributionMargin: 1500,
  },
};

describe('Product Computation Tests', () => {
  describe('getComputationResult', () => {
    it.each([
      ['basicProduct', mockData.basicProduct, expectedResults.basicProduct],
      [
        'complexProduct',
        mockData.complexProduct,
        expectedResults.complexProduct,
      ],
      [
        'edgeCaseProduct',
        mockData.edgeCaseProduct,
        expectedResults.edgeCaseProduct,
      ],
    ])(
      'should return correct computation result for %s',
      (_, product, expected) => {
        const result = getComputationResult(product as ProductSchema);
        expect(result).toEqual(expected);
      }
    );
  });

  describe('calculateManufacturingCostPerUnit', () => {
    it.each([
      [
        mockData.basicProduct.rawMaterials,
        mockData.basicProduct.overheadExpenses,
        1.5,
      ],
      [
        mockData.complexProduct.rawMaterials,
        mockData.complexProduct.overheadExpenses,
        1.5,
      ],
      [[], [], 0],
      [[{materialName: 'Test', totalUnits: 1, costPerUnit: 100}], [], 1],
      [[], [{expenseName: 'Test', totalUnits: 1, costPerUnit: 100}], 1],
    ])(
      'should calculate manufacturing cost correctly',
      (rawMaterials, overheadExpenses, expected) => {
        const result = productComputations.calculateManufacturingCostPerUnit(
          rawMaterials,
          overheadExpenses
        );
        expect(result).toBe(expected);
      }
    );
  });

  describe('calculateRecommendedRetailPrice', () => {
    it.each([
      [1.5, {calculationType: 'cost-plus' as const, profitValue: 20}, 1.8],
      [1.5, {calculationType: 'fixed-markup' as const, profitValue: 0.3}, 1.8],
      [100, {calculationType: 'cost-plus' as const, profitValue: 50}, 150],
      [100, {calculationType: 'fixed-markup' as const, profitValue: 50}, 150],
    ])(
      'should calculate retail price correctly',
      (cost, settings, expected) => {
        const result = productComputations.calculateRecommendedRetailPrice(
          cost,
          settings
        );
        expect(result).toBe(expected);
      }
    );
  });

  describe('calculateBreakEvenPoint', () => {
    it('should calculate break-even point correctly', () => {
      const result = productComputations.calculateBreakEvenPoint(
        mockData.basicProduct.overheadExpenses,
        mockData.basicProduct.rawMaterials,
        1.8
      );
      expect(result).toBe(62.5);
    });

    it('should throw an error when contribution margin is zero', () => {
      expect(() =>
        productComputations.calculateBreakEvenPoint(
          [{expenseName: 'Test', totalUnits: 1, costPerUnit: 1}],
          [{materialName: 'Test', totalUnits: 1, costPerUnit: 1}],
          1
        )
      ).toThrow();
    });
  });

  describe('calculateProfitPerUnit', () => {
    it.each([
      [1.8, 1.5, 0.3],
      [2, 1, 1],
      [1, 2, -1],
    ])(
      'should calculate profit per unit correctly',
      (price, cost, expected) => {
        const result = productComputations.calculateProfitPerUnit(price, cost);
        expect(result).toBe(expected);
      }
    );
  });

  describe('calculateTotalPotentialProfit', () => {
    it.each([
      [0.3, 80, 24],
      [1, 100, 100],
      [-0.5, 50, -25],
    ])(
      'should calculate total potential profit correctly',
      (profit, quantity, expected) => {
        const result = productComputations.calculateTotalPotentialProfit(
          profit,
          quantity
        );
        expect(result).toBe(expected);
      }
    );
  });

  describe('calculateMarginOfSafety', () => {
    it.each([
      [80, 62.5, 21.88],
      [100, 50, 50],
      [50, 50, 0],
      [0, 50, 0],
    ])(
      'should calculate margin of safety correctly',
      (sales, breakEven, expected) => {
        const result = productComputations.calculateMarginOfSafety(
          sales,
          breakEven
        );
        expect(result).toBeCloseTo(expected, 2);
      }
    );
  });

  describe('calculateContributionMargin', () => {
    it.each([
      [1.8, mockData.basicProduct.rawMaterials, 0.8],
      [2, [{materialName: 'Test', totalUnits: 1, costPerUnit: 1}], 1],
      [1, [{materialName: 'Test', totalUnits: 1, costPerUnit: 2}], -1],
    ])(
      'should calculate contribution margin correctly',
      (price, materials, expected) => {
        const result = productComputations.calculateContributionMargin(
          price,
          materials
        );
        expect(result).toBe(expected);
      }
    );
  });

  describe('Edge cases and error handling', () => {
    it.each([
      [Number.MAX_SAFE_INTEGER, 'should handle very large numbers'],
      [Number.MIN_VALUE, 'should handle very small numbers'],
      [-1, 'should handle negative numbers'],
    ])('calculateManufacturingCostPerUnit %s', input => {
      const result = productComputations.calculateManufacturingCostPerUnit(
        [{materialName: 'Test', totalUnits: input, costPerUnit: 1}],
        [{expenseName: 'Test', totalUnits: input, costPerUnit: 1}]
      );
      expect(result).toBeGreaterThan(0);
    });
  });

  describe('Decimal precision', () => {
    it('should maintain precision in chain calculations', () => {
      const result = getComputationResult({
        ...mockData.basicProduct,
        rawMaterials: [
          {materialName: 'Test', totalUnits: 1, costPerUnit: 1 / 3},
        ],
        overheadExpenses: [
          {expenseName: 'Test', totalUnits: 1, costPerUnit: 1 / 3},
        ],
      });
      expect(result.manufacturingCostPerUnit).toBeCloseTo(2 / 3, 10);
    });
  });
});
