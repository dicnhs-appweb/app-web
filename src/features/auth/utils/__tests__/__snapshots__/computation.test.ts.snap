// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Computation Tests Realistic data sets should correctly calculate all values for a realistic product 1`] = `
{
  "breakEvenPoint": 0.57,
  "contributionMargin": 87.5,
  "manufacturingCostPerUnit": 150,
  "marginOfSafety": 99.93,
  "profitPerUnit": 37.5,
  "recommendedRetailPrice": 187.5,
  "totalPotentialProfit": 30000,
}
`;

exports[`Product Computation Tests getComputationResult should return correct computation result 1`] = `
{
  "breakEvenPoint": 0.63,
  "contributionMargin": 80,
  "manufacturingCostPerUnit": 150,
  "marginOfSafety": 99.21,
  "profitPerUnit": 30,
  "recommendedRetailPrice": 180,
  "totalPotentialProfit": 2400,
}
`;
